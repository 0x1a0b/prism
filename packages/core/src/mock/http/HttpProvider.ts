import { IHttpOperation } from '@stoplight/types';
import { IExampleGenerator } from '../generator/IExampleGenerator';
import { IMockProvider } from '../IMockProvider';
import { IMockResult } from '../IMockResult';
import { IHttpOperationConfig, IHttpResponse } from '@stoplight/prism-http/types';

export class HttpProvider
  implements IMockProvider<IHttpOperation, IHttpOperationConfig, IMockResult<IHttpResponse>> {
  constructor(private exampleGenerator: IExampleGenerator<any>) {}

  public async mock(
    resource: IHttpOperation,
    config: IHttpOperationConfig
  ): Promise<IMockResult<IHttpResponse>> {
    const response = resource.responses.find(r => r.code === config.code);
    if (!response) {
      throw new Error(`Response for status code '${config.code}' not found`);
    }

    const content = response.content.find(c => c.mediaType === config.mediaType);
    if (!content) {
      throw new Error(`Content for media type '${config.mediaType}' not found`);
    }

    let body;
    if (config.dynamic) {
      if (!content.schema) {
        throw new Error('Cannot generate response, schema is missing');
      }

      body = await this.exampleGenerator.generate(content.schema, content.mediaType);
    } else {
      const example = content.examples && content.examples.find(e => e.key === config.exampleKey);

      if (!example) {
        throw new Error(`Example for key '${config.exampleKey}' not found`);
      }

      body = example.value;
    }

    return {
      data: {
        statusCode: parseInt(response.code),
        // @todo: HttpHeaderParam[] ?
        headers: { 'Content-type': content.mediaType },
        body,
      },
    };
  }
}
