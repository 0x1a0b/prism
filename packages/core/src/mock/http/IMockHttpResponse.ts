export interface IMockHttpResponse {
  status: number;
  body?: any;
  headers?: object;
}
