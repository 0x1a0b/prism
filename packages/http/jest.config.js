const config = {
  ...require('../../jest.base'),
};

console.log(process.cwd(), config);

module.exports = config;
