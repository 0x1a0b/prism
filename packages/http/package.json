{
  "name": "@stoplight/prism-http",
  "version": "0.0.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "author": "Stoplight <support@stoplight.io>",
  "license": "UNLICENSED",
  "repository": {
    "type": "git",
    "url": "https://github.com/stoplightio/prism.git"
  },
  "files": [
    "/lib"
  ],
  "engines": {
    "node": ">=8"
  },
  "dependencies": {
    "@stoplight/prism-core": "0.0.0",
    "axios": "^0.18.0",
    "json-schema-faker": "^0.5.0-rc16"
  }
}
