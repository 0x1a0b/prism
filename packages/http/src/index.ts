import { factory, filesystemLoader } from '@stoplight/prism-core';
import { IFilesystemLoaderOpts } from '@stoplight/prism-core/types';
import { IHttpOperation } from '@stoplight/types';

import { forwarder } from './forwarder';
import { mocker } from './mocker';
import { router } from './router';
import { IHttpConfig, IHttpRequest, IHttpResponse } from './types';
import * as types from './types';
import { validator } from './validator';

export { types };

export const createInstance = factory<
  IHttpOperation,
  IHttpRequest,
  IHttpResponse,
  IHttpConfig,
  IFilesystemLoaderOpts
>({
  config: {
    mock: true,
  },
  loader: filesystemLoader,
  router,
  forwarder,
  validator,
  mocker,
});
