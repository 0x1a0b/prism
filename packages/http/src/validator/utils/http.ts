import { IHttpContent } from '@stoplight/types/http-spec';

import { IHttpNameValue } from '../../types';

export function getHeaderByName(headers: IHttpNameValue, name: string): string | undefined {
  const key = Object.keys(headers).find(n => n.toLowerCase() === name.toLowerCase());
  if (!key) return;

  return headers[key];
}

export function resolveContent(content: { [mediaType: string]: IHttpContent }, mediaType?: string) {
  return content[mediaType || '*'] || content['*'];
}
