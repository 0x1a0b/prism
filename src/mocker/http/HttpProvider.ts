import { IHttpOperation } from '@stoplight/types';
import { IMockProvider } from '../IMockProvider';
import { IMockResult } from '../IMockResult';
import { IHttpOperationOptions } from './IHttpOperationOptions';
import { IMockHttpResponse } from './IMockHttpResponse';

export class HttpProvider
  implements IMockProvider<IHttpOperation, IHttpOperationOptions, IMockResult<IMockHttpResponse>> {
  public async mock(
    operation: IHttpOperation,
    options: IHttpOperationOptions
  ): Promise<IMockResult<IMockHttpResponse>> {
    const response = operation.responses.find(r => r.code === options.status);
    if (!response) {
      throw new Error(`Response for status code '${options.status}' not found`);
    }

    const content = response.content.find(c => c.mediaType === options.contentType);
    if (!content) {
      throw new Error(`Content for media type '${options.contentType}' not found`);
    }

    const example = content.examples && content.examples.find(e => e.key === options.example);
    if (!example) {
      throw new Error(`Example for key '${options.example}' not found`);
    }

    return {
      data: {
        status: parseInt(response.code),
        // @todo: HttpHeaderParam[] ?
        headers: { 'Content-type': content.mediaType },
        body: example.value,
      },
    };
  }
}
