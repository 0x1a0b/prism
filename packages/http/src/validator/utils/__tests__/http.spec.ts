import { getHeaderByName, resolveContent } from '../http';

describe('getHeaderByName()', () => {
  describe('Content-type header is present', () => {
    it('returns Content-type header value', () => {
      expect(getHeaderByName({ 'Content-type': 'application/json' }, 'content-type')).toBe(
        'application/json'
      );
    });
  });

  describe('Content-type header is not present', () => {
    it('returns undefined', () => {
      expect(getHeaderByName({}, 'content-type')).toBeUndefined();
    });
  });
});

describe('resolveContent()', () => {
  describe('mediaType not provided', () => {
    it('returns default content', () => {
      expect(
        resolveContent({
          '*': { mediaType: '*' },
          'application/exists-son': { mediaType: 'application/exists-son' },
        })
      ).toMatchSnapshot();
    });
  });

  describe('mediaType provided', () => {
    describe('specific content exists', () => {
      it('returns specific content', () => {
        expect(
          resolveContent(
            {
              '*': { mediaType: '*' },
              'application/exists-son': { mediaType: 'application/exists-son' },
            },
            'application/exists-son'
          )
        ).toMatchSnapshot();
      });
    });

    describe('specific content does not exist', () => {
      it('returns default content', () => {
        expect(
          resolveContent(
            {
              '*': { mediaType: '*' },
              'application/exists-son': { mediaType: 'application/exists-son' },
            },
            'application/non-exists-son'
          )
        ).toMatchSnapshot();
      });
    });
  });
});
