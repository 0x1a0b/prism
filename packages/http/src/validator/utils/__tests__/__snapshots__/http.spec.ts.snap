// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveContent() mediaType not provided returns default content 1`] = `
Object {
  "mediaType": "*",
}
`;

exports[`resolveContent() mediaType provided specific content does not exist returns default content 1`] = `
Object {
  "mediaType": "*",
}
`;

exports[`resolveContent() mediaType provided specific content exists returns specific content 1`] = `
Object {
  "mediaType": "application/exists-son",
}
`;
